### ISSUE TITLE
Warning 'useAsyncData must return a value (it should not be undefined) (...)' triggered for server route returning `null` or `''`

### ISSUE THREAD

**Issue by DamianGlowala** [nuxt/nuxt#28920]
### Environment

n/a

### Reproduction

https://stackblitz.com/edit/github-9s3uba

### Describe the bug

Warning 'useAsyncData must return a value (it should not be undefined) or the request may be duplicated on the client side.' occurs when calling server route which returns `null` or `''`.

### Additional context

Mentioned in https://github.com/nuxt/nuxt/issues/28096#issuecomment-2291100566

**Comment by danielroe** [comment]
this improvement has landed in 4.x but not in 3.x branch (nor will it) - so possibly the warning message needs to be updated.

**Comment by felixgabler** [comment]
> this improvement has landed in 4.x but not in 3.x branch (nor will it) - so possibly the warning message needs to be updated.

Thanks @danielroe! However, I am still a bit confused and would be very grateful if you could confirm my understanding:

In Nuxt 4 (alpha) every value except `undefined` is now serialised into the SSR payload, so returning `null`, `''`, `0`, etc. from `useAsyncData` will not trigger a second client-side fetch. Nuxt 3 keeps the old behaviour, and the current warning text just needs updating. In other words, upgrading to Nuxt 4 is the only way to get “return null without refetch” – the compatibility flag in a Nuxt 3 project doesn’t change this (or I am perhaps doing something wrong). Is that correct?

**Comment by danielroe** [comment]
@felixgabler yes, that's correct.

**Comment by danielroe** [comment]
looking at the original issue, I'd just observe that returning `null` from a server endpoint actually is code for returning a 204 response (so in fact fetching from `/api/test` in this case _does_ return undefined).

I've updated the warning and made the console log more accurate.

**Comment by felixgabler** [comment]
> [@felixgabler](https://github.com/felixgabler) yes, that's correct.

@danielroe thanks for confirming! However, I just updated to Nuxt 4 and it still triggers a second client-side fetch when I return `null`. I have to use a stand in string literal `'null'` to make sure it does not fetch again. Did I misunderstand something? How is it meant to work to return a "non-existent" in `useAsyncData` that does not trigger another client-side fetch?

**Comment by danielroe** [comment]
are you returning null from a nitro endpoint? if so, you should be aware that actually returns 204 and an undefined body, rather than 200 with `null`

**Comment by felixgabler** [comment]
> are you returning null from a nitro endpoint? if so, you should be aware that actually returns 204 and an undefined body, rather than 200 with `null`

No, we return directly based on a variable being undefined.

Here is the entire context in our `2.exam.ts` plugin file:

```typescript
const {
    data: internalExam,
    status: examStatus,
    refresh: internalRefreshExam,
    // Nuxt 3 even in compatibility mode still refetches if value not truthy,
    // so we use 'null' instead of null to avoid this for now.
    // https://github.com/nuxt/nuxt/issues/28920
    // TODO: Remove this once we upgrade to Nuxt 4
} = useLazyAsyncData<ExamBasics | 'null'>(
    key,
    async () => {
        if (!unverifiedExamSlug.value) {
            return 'null';
        }

        ... supabase query construction

        const { data } = await query.maybeSingle();

        if (isNullable(data)) {
            return 'null';
        }

        return data;
    },
    { immediate: false, getCachedData },
);

// Track the fetch promise so middleware can await it before doing auth checks
const fetchingExamPromise = ref(Promise.resolve());

async function refreshExam() {
    if (examStatus.value === 'pending') {
        return fetchingExamPromise.value;
    }
    await (fetchingExamPromise.value = internalRefreshExam());
}

// Start the fetch on server-side only (client gets data from SSR cache)
// This prevents double-fetching while still allowing middleware to wait for completion
if (internalExam.value === undefined) refreshExam();

const exam = computed(() => (internalExam.value === 'null' ? undefined : (internalExam.value ?? undefined)));
```

Like this, with `'null'`, it does not trigger it again. But if we switch out `'null'` with `null`, it starts fetching multiple times. What are we doing incorrectly? Is it the fact that this is in a plugin?

**Comment by felixgabler** [comment]
@danielroe sorry to tag you directly but do you have any idea what is going wrong here?

**Comment by danielroe** [comment]
@felixgabler do you have a minimal reproduction?

**Comment by felixgabler** [comment]
> [@felixgabler](https://github.com/felixgabler) do you have a minimal reproduction?

Yes, I managed to repro it here: https://codesandbox.io/p/devbox/pedantic-platform-vf79w6?workspaceId=ws_36RKmjcjsYeB3TpkXzy8Nv

But on further inspection, I figured out that it was due to my implementation of `getCachedData`, which previously had:

```typescript
function getCachedData(key: string, nuxtApp: NuxtApp, ctx: { cause: string }) {
    // Always refetch on manual refresh
    if (ctx.cause === 'refresh:manual') {
        return;
    }

    // Use cache for initial loads, navigation, and watcher triggers
    return nuxtApp.payload.data[key] || nuxtApp.static.data[key];
}
```

But it should actually be:

```typescript
function getCachedData(key: string, nuxtApp: NuxtApp, ctx: { cause: string }) {
    // Always refetch on manual refresh
    if (ctx.cause === 'refresh:manual') {
        return;
    }

    // Use cache for initial loads, navigation, and watcher triggers (allow null)
    if (nuxtApp.payload.data[key] === undefined) {
        return nuxtApp.static.data[key];
    }

    return nuxtApp.payload.data[key];
}
```

I originally got the `getCachedData` from https://github.com/nuxt/nuxt/discussions/27006 but I think with the latest changes this is outdated. Happy that it works now!

**Comment by danielroe** [comment]
fantastic! 🙏
